cpp/src/arrow/util/parallel.h

```diff
@@ -35,61 +32,19 @@ namespace internal {
 template <class FUNCTION>
 Status ParallelFor(int num_tasks, FUNCTION&& func) {
   auto pool = internal::GetCpuThreadPool();
-  std::vector<std::future<Status>> futures(num_tasks);
+  std::vector<Future<Status>> futures;
+  futures.reserve(num_tasks);
```

Comment:
@bkietz Note that a Future isn't default-constructible. Should I add a default constructor that initializes it to an invalid state?

Reply:
My concern is: `Future::Make` does some allocations. If we allow writing:
```c++
std::vector<std::future<Status>> futures(num_tasks);
```
then we're allocating some future internals, only to replace them with other internals when assigning the vector elements. Which is more wasteful than what the code suggests.